pygmt/clib/session.py

```diff
@@ -1653,6 +1656,66 @@ def virtualfile_from_grid(self, grid: xr.DataArray) -> Generator[str, None, None
         ) as vfile:
             yield vfile
 
+    @contextlib.contextmanager
+    def virtualfile_from_image(self, image: xr.DataArray):
+        """
+        Store an image in a virtual file.
+
+        Use the virtual file name to pass in the data in your image to a GMT module.
+        Images must be :class:`xarray.DataArray` instances.
+
+        Context manager (use in a ``with`` block). Yields the virtual file name that you
+        can pass as an argument to a GMT module call. Closes the virtual file upon exit
+        of the ``with`` block.
+
+        The virtual file will contain the image as a ``GMT_MATRIX`` with extra metadata.
+
+        Use this instead of creating a data container and virtual file by hand with
+        :meth:`pygmt.clib.Session.create_data`, :meth:`pygmt.clib.Session.put_matrix`,
+        and :meth:`pygmt.clib.Session.open_virtualfile`.
+
+        The image data matrix must be C contiguous in memory. If it is not (e.g., it is
+        a slice of a larger array), the array will be copied to make sure it is.
+
+        Parameters
+        ----------
+        image : :class:`xarray.DataArray`
+            The image that will be included in the virtual file.
+
+        Yields
+        ------
+        fname : str
+            The name of virtual file. Pass this as a file name argument to a GMT module.
+
+        """
+        _gtype = {0: "GMT_GRID_IS_CARTESIAN", 1: "GMT_GRID_IS_GEO"}[image.gmt.gtype]
+        _reg = {0: "GMT_GRID_NODE_REG", 1: "GMT_GRID_PIXEL_REG"}[image.gmt.registration]
+
+        # Conversion to a C-contiguous array needs to be done here and not in put_matrix
+        # because we need to maintain a reference to the copy while it is being used by
+        # the C API. Otherwise, the array would be garbage collected and the memory
+        # freed. Creating it in this context manager guarantees that the copy will be
+        # around until the virtual file is closed. The conversion is implicit in
+        # dataarray_to_matrix.
+        nbands, nrows, ncols = image.shape  # CHW
+        matrix, _region, _inc = dataarray_to_matrix(image)
+
+        family = "GMT_IS_IMAGE|GMT_VIA_MATRIX"
+        geometry = "GMT_IS_SURFACE"
+        gmt_image = self.create_data(
+            family,
+            geometry,
+            mode=f"GMT_CONTAINER_ONLY|{_gtype}",
+            dim=(ncols, nrows, nbands, self["GMT_UCHAR"]),
```

Comment:
@seisman, can you check if I'm passing the `dim` arguments correctly to `create_data`? Have looked at your notes added in https://github.com/GenericMappingTools/pygmt/pull/3532, and at https://docs.generic-mapping-tools.org/6.5/devdocs/api.html#create-empty-resources that it should be a tuple of 1) number of columns, 2) number of rows, 3) number of layers(bands?), and 4) data type of matrix (UCHAR?). But still getting a segfault.